### ISSUE TITLE
Forever with `npm start`

### ISSUE THREAD

**Issue by electerious** [foreversd/forever#540]
Hi,

I'm normally starting my script with `npm start` to get access to the user environment, which includes variables from the package.json (e.g. `process.env.npm_package_config_port`).

Is there a way to start my script with `npm start` in forever? Or an alternative way to get the npm environment variables when starting the script with forever?

**Comment by ambodi** [comment]
How do you actually stop the process this way?

**Comment by danawoodman** [comment]
+1 for actually killing the script when using npm tasks. `forever stop <PID>` doesn't actually work in killing the processes spawned by the task.

**Comment by woodyrew** [comment]
I have the same issue, am I using `forever` wrongly?
```
$ forever --version
v0.15.1

$ forever start -c "npm start" /some/dir
...bash
info:    Forever processing file: /some/dir/

$ forever list
info:    Forever processes running
data:        uid  command   script     forever pid  id logfile                    uptime      
data:    [0] 9T8K npm start /some/dir/ 8031    8048    /home/me/.forever/9T8K.log 0:0:0:9.371 

$ forever stop 0
info:    Forever stopped process:
    uid  command   script     forever pid  id logfile                    uptime       
[0] 9T8K npm start /some/dir/ 8031    8048    /home/me/.forever/9T8K.log 0:0:0:23.884 

$ forever list
info:    No forever processes running

$ ps aux | grep 'forever\|node'
me     8065  0.0  0.0   4444   660 ?        S    14:00   0:00 sh -c NODE_PATH=$NODE_PATH:./modules node .
me     8066  0.0  0.4 705192 37700 ?        Sl   14:00   0:00 node .
me     8531  0.0  0.0  15944   944 pts/27   S+   14:14   0:00 grep --color=auto forever\|node
```

`ps kill 8065` leaves `8066` running and script is still responding via http.
`ps kill 8066` removes both processes and brings down the http server.

Forever lives up to its name! :wink: I hope this information is useful.

**Comment by dineshsprabu** [comment]
Solution for NPM START problem with forever.

https://gist.github.com/dineshsprabu/e7553cc9cc2f23b5c696

**Comment by oakinogundeji** [comment]
Hi Guys,

I had this same issue and it was driving me crazy, the solution was mind boggling in it's simplicity.
Initially, my package.json had the following scripts:
```
"scripts": {
    "test": "mocha test",
    "start": "npm run build && node server.js",
    "start-dev": "nodemon server.js",
    "build": "browserify -t partialify ./public/js/app.js -o ./public/js/main.js",
    "watch": "watchify -t partialify ./public/js/app.js -o ./public/js/main.js -dv"
  },
```
I had tried 
`forever start -c "npm start" /path/to/app/dir/` 
because i couldn't get it to work with 'upstart', the solution was to do the following:
`"start": "npm run build && forever start --minUptime 1000 --spinSleepTime 1000 server.js",`
and then from the root of my app simply run 'npm start' and BAM.... it's working.
Try it and let me know how it worked.

Regards

**Comment by Ankur-Marwaha** [comment]
This is what I am trying to do and it is not working and By the way what is server.js over here.

    "start": "node ./bin/www && forever start --minUptime 1000 --spinSleepTime 1000 server.js"

**Comment by wayofthefuture** [comment]
Any simple solution for this yet?

**Comment by oakinogundeji** [comment]
@Ankur-Marwaha sorry i've taken soo long to respond. server.js is the entry point to the app. How did it go for you? Were you successful? You need to run 'npm start' from the project root.

Of course this is no substitute for properly daemonizing the app, but nevertheless it does serve.

**Comment by malhotraa** [comment]
Hi 

I would like to to detect my deployment configs specified in the config/ folder of my application root directory. Does anyone know how to do this with forever? Currently it is ignoring the config specified in that folder and my application is complaining that several config variables arent defined.

Thanks

**Comment by TheMaverickProgrammer** [comment]
This doesn't work with the `nohup ... &` command. Often I need to read logs from my react page.

**Comment by nccho** [comment]
How to use on Windows ?
forever start -c "npm start" ???

**Comment by knutole** [comment]
### Clarification regarding `forever` _daemon_ mode

If you're using Docker and your container exists, it's possibly because your `forever` is running as a daemon. Since all scripts have finished, your Docker container will exit. The solution is to use non-daemon mode of `forever`:

#### Daemon mode (will exit your Docker container):
```bash
# good bye, container
forever start ./bin/www
```

#### Non-daemon mode (will not exit):
```bash
# try this instead
forever ./bin/www
```

This should work both in a bash script (above syntax) and in package.json:
```json
// package.json
"start" : "forever ./bin/www"
```

**Comment by CodeWithOz** [comment]
Hey everyone,

**tl;dr**:
If you have a path in the command you supply to the `-c` option, move that path to the position where forever expects a script/path. Forever passes that path to your command, so placing the path in the command will cause 2 path arguments to be supplied to your command.

---

**Longer version**

I had a similar issue when using `serve`. My command was `npm run serve` and my `package.json` had:
```
...
"serve": "npx serve -s build/"
...
```

So I was wrongly using:
```
forever start -c "npm run serve" ./
```

The problem is that forever passes the path I provided (`./`) to the command I provided (`npm run serve`), so ultimately the command became:
```
npx serve -s build/ /home/username/path/to/app
```
This was throwing an error saying
>`ERROR: Please provide one path argument at maximum`.

That's how `serve` complains about receiving two paths instead of only one.
My solution was to change my `package.json` to:
```
...
"serve": "npx serve -s"
...
```

and then supply the `build/` folder through forever, like this:
```
forever start -c "npm run serve" build/
```
And now it works :+1: .

**Comment by milad1367** [comment]
I used this ```forever start -c "npm start" ./```  worked fine!

**Comment by Fahdshehzad** [comment]
This works: `forever start ./bin/www`

**Comment by manoharreddyporeddy** [comment]
Feb 2021

This solution is for express.js & forever.

package.json

```
  "scripts": {
    "start": "set PORT=3001 && set DEBUG=myapp:* && forever start --minUptime 1000 --spinSleepTime 1000 ./bin/www",
  }
```

Now run:

    npm start

NOTE: Above works on Windows OS, the && is required, for others it will be different, see following link for other OS: https://expressjs.com/en/starter/generator.html

**Comment by duck-moon9392** [comment]
OMG. It works! Thank you so much.
